src/NHibernate/Context/MapBasedSessionContext.cs

```diff
@@ -20,30 +23,35 @@ protected override ISession Session
 		{
 			get
 			{
-				IDictionary map = GetMap();
-				if (map == null)
-				{
-					return null;
-				}
-				else
-				{
-					return map[_factory] as ISession;
-				}
+				return GetConcreteMap()?.GetOrAdd(_factory, default(ISession));
 			}
 			set
 			{
-				IDictionary map = GetMap();
+				var map = GetConcreteMap();
 				if (map == null)
 				{
-					map = new Hashtable();
-					SetMap(map);
+					// Double check locking. Cannot use a Lazy<T> for such a class semantic.
+					lock (_locker)
+					{
+						map = GetConcreteMap();
+						if (map == null)
+						{
+							map = new ConcurrentDictionary<ISessionFactoryImplementor, ISession>();
+							SetMap(map);
+						}
+					}
 				}
 				map[_factory] = value;
 			}
 		}
 
+		private ConcurrentDictionary<ISessionFactoryImplementor, ISession> GetConcreteMap()
+		{
+			return (ConcurrentDictionary<ISessionFactoryImplementor, ISession>)GetMap();
+		}
+
 		/// <summary>
-		/// Get the dictionary mapping session factory to its current session.
+		/// Get the dictionary mapping session factory to its current session. Yield <c>null</c> if none have been set.
 		/// </summary>
 		protected abstract IDictionary GetMap();
```

Comment:
I could have changed the returned type for the concrete class type in order to avoid the above cast, but this would be a breaking change for external implementers, and this would expose a technical detail.